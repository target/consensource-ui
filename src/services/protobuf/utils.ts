import {
  CertificateRegistryPayload,
  ICertificateRegistryPayload,
} from './compiled';

export interface PayloadInfo {
  payloadBytes: string | Buffer | NodeJS.TypedArray | DataView;
  inputs: string[];
  outputs: string[];
}

export const ACTIONS = CertificateRegistryPayload.Action;

/**
 * Encodes a CertificateRegistryPayload message
 * @param message CertificateRegistryPayload message or plain object to encode
 */
export function encodePayload(
  payload: ICertificateRegistryPayload,
): Uint8Array {
  return CertificateRegistryPayload.encode(payload).finish();
}
