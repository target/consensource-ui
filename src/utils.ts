// Util collection

/**
 * Selects keys from an object
 */
export const pluck = (obj: any, ...keys: Array<any>): any =>
    keys.reduce((o, k) => {
        if (typeof obj[k] !== 'undefined') {
            o[k] = obj[k];
        }
        return o;
    }, {});

/**
 * Merges objects, starting from left to right
 */
