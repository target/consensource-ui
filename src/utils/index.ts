export function getLocaleFromUnix(timestamp: number) {
  return new Date(timestamp).toLocaleDateString();
}
