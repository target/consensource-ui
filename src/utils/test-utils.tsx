import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { StoresContext } from 'view/context';
import { stores } from 'stores';
import { theme } from 'config/theme';
import { ThemeProvider } from '@material-ui/core';
import CssBaseline from '@material-ui/core/CssBaseline';

const wrapper = ({ children }: any) => {
  return (
    <CssBaseline>
      <ThemeProvider theme={theme}>
        <StoresContext.Provider value={stores}>
          <MemoryRouter initialEntries={['/']}>{children}</MemoryRouter>
        </StoresContext.Provider>
      </ThemeProvider>
    </CssBaseline>
  );
};

const customRender = (ui: any, options?: any) =>
  render(ui, { wrapper, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
