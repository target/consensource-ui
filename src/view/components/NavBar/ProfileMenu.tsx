import React, { useState } from 'react';
import { AccountCircle as ProfileIcon } from '@material-ui/icons';
import {
  Menu,
  MenuItem,
  IconButton,
  IconButtonProps,
  Typography,
  Grid,
} from '@material-ui/core';
import { useHistory } from 'react-router-dom';
import { useStores } from 'services/hooks';
import { observer } from 'mobx-react-lite';

export interface NavbarProfileIconProps {
  onClick: IconButtonProps['onClick'];
  username?: string;
}

export const NavbarProfileIcon = ({
  onClick,
  username,
}: NavbarProfileIconProps) => {
  return (
    <IconButton
      color="inherit"
      aria-label="profile"
      onClick={onClick}
      edge="start"
      data-testid="profile-icon-button"
    >
      <Grid container alignItems="center" spacing={1}>
        <ProfileIcon />

        {username && (
          <Grid item>
            <Typography data-testid="username">{username}</Typography>
          </Grid>
        )}
      </Grid>
    </IconButton>
  );
};

export const ProfileMenu = observer(() => {
  const { userStore } = useStores();
  const history = useHistory();
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const username = userStore?.user?.username;

  const handleLogout = () => {
    userStore.logout();
    history.push('/');
  };

  const handleProfileClick = () => {
    setAnchorEl(null);
    history.push('/profile');
  };

  return (
    <div>
      <NavbarProfileIcon
        onClick={(e) => setAnchorEl(e.currentTarget)}
        username={username}
      />
      <Menu
        keepMounted
        anchorEl={anchorEl}
        open={!!anchorEl}
        onClose={() => setAnchorEl(null)}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'center',
        }}
        getContentAnchorEl={null}
      >
        <MenuItem onClick={handleProfileClick}>Profile</MenuItem>
        <MenuItem onClick={handleLogout}>Logout</MenuItem>
      </Menu>
    </div>
  );
});
