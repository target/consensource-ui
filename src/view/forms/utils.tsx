import React from 'react';
import { Grid, Typography, TextFieldProps } from '@material-ui/core';
import { BatchStatusRes } from 'services/api';

export interface TransactionFormProps {
  setBatchStatusLink: (statusLink: BatchStatusRes['link']) => void;
}

export type onChangeEvent = Parameters<
  NonNullable<TextFieldProps['onChange']>
>[0];

/**
 * Helper function that checks if any values in `state` have not been set.
 * Returns true if there is any key in `state` that is falsey.
 */
export function hasEmptyFields<T>(state: T) {
  return Object.keys(state).some((key) => !state[key as keyof T]);
}

export interface FormErrMsgProps {
  msg?: string;
}

export function FormErrMsg({ msg }: FormErrMsgProps) {
  if (!msg) {
    return null;
  }

  return (
    <Grid item>
      <Typography variant="caption" color="error">
        {msg}
      </Typography>
    </Grid>
  );
}
