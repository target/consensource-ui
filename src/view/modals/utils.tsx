import { DialogProps as MUIDialogProps } from '@material-ui/core';
import { FactoryResData } from 'services/api';

export interface DialogProps extends MUIDialogProps {
  handleClose: () => void;
}

export interface OrgDialogProps extends DialogProps {
  existing_org: FactoryResData;
}
