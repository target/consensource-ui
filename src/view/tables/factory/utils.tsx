import { MUIDataTableOptions, MUIDataTableChip } from 'mui-datatables';

export const DEFAULT_ROWS_PER_PAGE = 100;

export const textLabels: MUIDataTableOptions['textLabels'] = {
  body: {
    noMatch: 'No factories found',
  },
};

/**
 * Default props for the filter chips displayed in the
 * `<TableFilterList />`.
 */
export const filterChipProps: MUIDataTableChip = {
  color: 'secondary',
  variant: 'default',
};

/**
 * Utility function to convert a string to an array, or return
 * the original param if an array was passed.
 *
 * This operation is frequent when decoding URL query params.
 */
export const convertStrToArray = (val: string[] | string) => {
  return typeof val === 'string' ? [val] : val;
};
