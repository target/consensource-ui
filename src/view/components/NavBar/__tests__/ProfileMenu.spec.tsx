import React from 'react';
import { render, fireEvent, screen } from 'utils/test-utils';
import { stores } from 'stores';
import { ProfileMenu } from '../ProfileMenu';

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

describe('<ProfileMenu />', () => {
  it('opens the profile menu on click', () => {
    const { container } = render(<ProfileMenu />);

    fireEvent.click(screen.getByRole('button'));

    expect(screen.getByRole('menu')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('renders without a username', () => {
    const { container } = render(<ProfileMenu />);
    expect(screen.queryByTestId('username')).toBeNull();
    expect(container).toMatchSnapshot();
  });

  it('renders the username', () => {
    const username = 'Bob';

    jest
      .spyOn(stores.userStore, 'user', 'get')
      .mockReturnValueOnce({ username } as any);

    const { container } = render(<ProfileMenu />);

    expect(screen.getByText(username)).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('logs a user out and redirects to "/"', () => {
    render(<ProfileMenu />);

    const spy = jest.spyOn(stores.userStore, 'logout');

    fireEvent.click(screen.getByRole('button'));
    fireEvent.click(screen.getByText('Logout'));

    expect(spy).toHaveBeenCalled();
    expect(mockHistoryPush).toHaveBeenCalledWith('/');
  });

  it('redirects a user to "/profile"', () => {
    render(<ProfileMenu />);

    fireEvent.click(screen.getByRole('button'));
    fireEvent.click(screen.getByText('Profile'));

    expect(mockHistoryPush).toHaveBeenCalledWith('/profile');
  });
});
