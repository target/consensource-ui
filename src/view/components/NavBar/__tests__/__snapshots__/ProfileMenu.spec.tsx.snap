// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProfileMenu /> opens the profile menu on click 1`] = `
<div
  aria-hidden="true"
>
  <div>
    <button
      aria-label="profile"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-align-items-xs-center"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
            />
          </svg>
        </div>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`<ProfileMenu /> renders the username 1`] = `
<div>
  <div>
    <button
      aria-label="profile"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-align-items-xs-center"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
            />
          </svg>
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              Bob
            </p>
          </div>
        </div>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`<ProfileMenu /> renders without a username 1`] = `
<div>
  <div>
    <button
      aria-label="profile"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-align-items-xs-center"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
            />
          </svg>
        </div>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
